pkg/cmd/release/download/download.go

```
@@ -0,0 +1,171 @@
+			}
+			return downloadRun(opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.Destination, "dir", "C", ".", "The directory to download files into")
+
+	return cmd
+}
+
+func downloadRun(opts *DownloadOptions) error {
+	httpClient, err := opts.HttpClient()
+	if err != nil {
+		return err
+	}
+
+	baseRepo, err := opts.BaseRepo()
+	if err != nil {
+		return err
+	}
+
+	release, err := shared.FetchRelease(httpClient, baseRepo, opts.TagName)
+	if err != nil {
+		return err
+	}
+
+	var toDownload []shared.ReleaseAsset
+	for _, a := range release.Assets {
+		if opts.FilePattern != "" {
+			if isMatch, err := filepath.Match(opts.FilePattern, a.Name); err != nil || !isMatch {
+				continue
+			}
+		}
+		toDownload = append(toDownload, a)
+	}
+
+	if len(toDownload) == 0 {
+		if len(release.Assets) > 0 {
+			return errors.New("no assets match the file pattern")
+		}
+		return errors.New("no assets to download")
+	}
+
+	if opts.Destination != "." {
+		err := os.MkdirAll(opts.Destination, 0755)
+		if err != nil {
+			return err
+		}
+	}
+
+	opts.IO.StartProgressIndicator()
+	err = downloadAssets(httpClient, toDownload, opts.Destination, opts.Concurrency)
+	opts.IO.StopProgressIndicator()
+	return err
+}
+
+func downloadAssets(httpClient *http.Client, toDownload []shared.ReleaseAsset, destDir string, numWorkers int) error {
+	if numWorkers == 0 {
+		return errors.New("the number of concurrent workers needs to be greater than 0")
+	}
+
+	jobs := make(chan shared.ReleaseAsset, len(toDownload))
+	results := make(chan error, len(toDownload))
+
+	if len(toDownload) < numWorkers {
+		numWorkers = len(toDownload)
+	}
+
+	for w := 1; w <= numWorkers; w++ {
+		go func() {
+			for a := range jobs {
+				results <- downloadAsset(httpClient, a.APIURL, filepath.Join(destDir, a.Name))
+			}
+		}()
+	}
+
+	for _, a := range toDownload {
+		jobs <- a
+	}
+	close(jobs)
+
+	var downloadError error
+	for i := 0; i < len(toDownload); i++ {
+		if err := <-results; err != nil {
+			downloadError = err
```

I wonder if it's worth a `break` here if we're not going to collect errors per-download. Otherwise we're discarding errors for all but the last download which is confusing.

---

Good thought, but we cannot `break` here since we need to invoke `<-results` as many times as there were download jobs. You're right that all but the latest error gets discarded. Right now I don't think that's such an issue, but if people notice this, we could collect all errors into one meta-error